core/src/main/java/org/apache/calcite/sql/fun/SqlExtractFunction.java

```diff
@@ -83,8 +88,102 @@ public SqlExtractFunction(String name) {
     //    startUnit = EPOCH and timeFrameName = 'MINUTE15'.
     //
     // If the latter, check that timeFrameName is valid.
-    validator.validateTimeFrame(
-        (SqlIntervalQualifier) call.getOperandList().get(0));
+    SqlIntervalQualifier qualifier = call.operand(0);
+    validator.validateTimeFrame(qualifier);
+    TimeUnitRange range = qualifier.timeUnitRange;
+
+    RelDataType type = validator.getValidatedNodeTypeIfKnown(call.operand(1));
+    if (type == null) {
+      return;
+    }
+
+    SqlTypeName typeName = type.getSqlTypeName();
+    boolean legal;
+    switch (range) {
+    case YEAR:
+    case MONTH:
+    case ISOYEAR:
+    case QUARTER:
+    case DECADE:
+    case CENTURY:
+    case MILLENNIUM:
+      legal = new ImmutableList.Builder<SqlTypeName>()
+          .add(SqlTypeName.DATE)
+          .add(SqlTypeName.TIMESTAMP)
+          .add(SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE)
+          .addAll(SqlTypeName.YEAR_INTERVAL_TYPES)
+          .build()
+          .contains(typeName);
+      break;
+    case WEEK:
+    case DOW:
+    case ISODOW:
+    case DOY:
+      legal = new ImmutableList.Builder<SqlTypeName>()
```

Comment:
should we extract the result of the builder here and others into constants (may be sets instead of lists) to avoid building it each time we come here?

Reply:
I am now using static fields for these constant lists.

Reply:
Why not using sets as @snuyanzin suggested?